### ISSUE TITLE
Error when pulling data

### ISSUE THREAD

**Issue by BastienFR** [gavinsimpson/canadaHCD#27]
It seems the package is broken from what I suspect being a change on EC website.  Now, when we try the example on the main page, we get :

```
library("canadaHCD")
find_station("Yellowknife")
# A tibble: 7 x 5
  Name                  Province              StationID LatitudeDD LongitudeDD
  <fct>                 <fct>                 <fct>          <dbl>       <dbl>
1 YELLOWKNIFE A         Northwest Territories 1706            62.5       -114.
2 YELLOWKNIFE A         Northwest Territories 51058           62.5       -114.
3 YELLOWKNIFE OVERLAP   Northwest Territories 53898           62.5       -114.
4 YELLOWKNIFE-HENDERSON Northwest Territories 45467           62.4       -114.
5 YELLOWKNIFE CON MINE  Northwest Territories 8949            62.5       -114.
6 YELLOWKNIFE CS        Northwest Territories 27338           62.5       -114.
7 YELLOWKNIFE HYDRO     Northwest Territories 1707            62.7       -114.
yh <- hcd_monthly(1707)
  |                                                                                                                                          |   0%Parsed with column specification:
cols(
  .default = col_double(),
  `Station Name` = col_character(),
  `Date/Time` = col_character(),
  Month = col_character(),
  `Mean Max Temp Flag` = col_character(),
  `Mean Min Temp Flag` = col_character(),
  `Mean Temp Flag` = col_character(),
  `Extr Max Temp Flag` = col_character(),
  `Extr Min Temp Flag` = col_character(),
  `Total Rain Flag` = col_character(),
  `Total Snow Flag` = col_character(),
  `Total Precip Flag` = col_character(),
  `Snow Grnd Last Day Flag` = col_character(),
  `Dir of Max Gust (10's deg)` = col_logical(),
  `Dir of Max Gust Flag` = col_logical(),
  `Spd of Max Gust (km/h)` = col_logical(),
  `Spd of Max Gust Flag` = col_logical()
)
See spec(...) for full column specifications.
Error: Columns `13`, `14`, `15`, `16` cannot have NA as name
```
Any idea what the problem might be?
Thanks

**Comment by ConorIA** [comment]
It looks like the ECCC CSV files no longer include the header with metadata, which means that the use of SKIP to choose which type of data we are downloading is no longer valid. Changes need to be made to the `read_hcd()` function.

**Comment by BastienFR** [comment]
I see.  Do you have a hourly, daily and monthly old version of the files that used to work?  Even if I'm not super confident of being able to fix that, I'll give it a try.  It would be easier to see what worked and trouble shot if I have a version of the old files produced by EC.

**Comment by gavinsimpson** [comment]
They've also changed the format of the files; there are now 4 additional columns (lat/Long, station name, station id). I should be able to fix this reasonably quickly; we don't need to skip anything; I just need to work out if I can continue to distinguish between the types of data on the basis of how many cols there are.

This all seems like a futile game of whack-a-mole; these data are available via a web service and I have code that reads those data from JSON which is documented, which would be preferable to having to continually fix up these incessant file format changes from Env Can.

**Comment by BastienFR** [comment]
Even if you assign it to yourself, know that I worked on it and modified the function `read_hcd`.  

```
`read_hcd` <- function(file, flags = FALSE, clean = TRUE, ...) {
    Date <- NULL                        # to kill the global variable NOTE!
    #fields <- read.table(file, nrows=1, header = T, sep=",", check.names = F)
    fields <- colnames(read_csv(file, n_max=1))
    if(any(fields=="Time")) types  <- "nncnTiiicdcdcicicicdcdcicicc" else {
        if(any(fields=="Day")) types  <- paste0("nncnDiiic", paste0(rep("dc", 8L), collapse = ""), "iciccc") else {
            types  <- paste0("nncncii", paste0(rep("dc", 8L), collapse = ""), "iciccc") 
        } 
    }
    df <- read_csv(file, locale = locale(encoding = "ISO-8859-1"),
                   col_types = types, ...)
    df <- as_data_frame(df)
    ## drop Year Month Day Time columns if exist
    df <- df[, !names(df) %in% c("Year", "Month", "Day", "Time")]

    ## clean variables names
    if (isTRUE(clean)) {
        names(df) <-
            if (inherits(df[[1]], "Date")) { # must be daily
                c("Longitude","Latitude","StationName","ClimateID",
                  "Date","Data Quality",
                  "MaxTemp","MaxTempFlag","MinTemp","MinTempFlag","MeanTemp",
                  "MeanTempFlag","HeatDegDays","HeatDegDaysFlag","CoolDegDays",
                  "CoolDegDaysFlag","TotalRain","TotalRainFlag","TotalSnow",
                  "TotalSnowFlag","TotalPrecip","TotalPrecipFlag","GroundSnow",
                  "GroundSnowFlag","MaxGustDir","MaxGustDirFlag","MaxGustSpeed",
                  "MaxGustSpeedFlag")
            } else if (inherits(df[[1]], "POSIXt")) { # must be hourly
                c("Longitude","Latitude","StationName","ClimateID",
                  "DateTime","Temp","TempFlag","DewPointTemp","DewPointTempFlag",
                  "RelHumidity","RelHumidityFlag","WindDir","WindDirFlag",
                  "WindSpeed","WindSpeedFlag","Visibility","VisibilityFlag",
                  "Pressure","PressureFlag","Humidex","HumidexFlag","WindChill",
                  "WindChillFlag","Weather")
            } else {                    # must be monthly
                c("Longitude","Latitude","StationName","ClimateID",
                  "Date","MaxTemp","MaxTempFlag","MinTemp","MinTempFlag",
                  "MeanTemp","MeanTempFlag","ExtremeHigh","ExtremeHighFlag","ExtremeLow",
                  "ExtremeLowFlag","TotalRain","TotalRainFlag","TotalSnow",
                  "TotalSnowFlag","TotalPrecip","TotalPrecipFlag","LastSnowGrnd",
                  "LastSnowGrndFlag","MaxGustDir","MaxGustDirFlag","MaxGustSpeed",
                  "MaxGustSpeedFlag")
            }
    }

    ## keep flag columns?
    flagCols <- grepl("Flag", names(df))
    if (!isTRUE(flags)) {
        df <- df[, !flagCols]
        df <- df[, !grepl("Data Quality", names(df))]
    }

    if (!(inherits(df[[5]], "Date") || inherits(df[[1]], "POSIXt"))) {
        ## coerce Date/Time to Zoo yearmon object
        df <- mutate(df, Date = as.yearmon(Date, format = "%Y-%m"))
    }

    ## return
    df
}
```
This fixes the problem, but not for the hourly data where often data are missing in the source data (lines are incomplete) which make the reading fail.  

```
hr <- hcd_hourly(station = 51157, year=2017, month = 3)
  |                                                                                                                                              |   0%Parsed with column specification:
cols(
  .default = col_logical(),
  `Longitude (x)` = col_double(),
  `Latitude (y)` = col_double(),
  `Station Name` = col_character(),
  `Climate ID` = col_double(),
  `Date/Time` = col_datetime(format = ""),
  Year = col_double(),
  Month = col_character(),
  Day = col_character(),
  Time = col_time(format = ""),
  `Temp (°C)` = col_double(),
  `Dew Point Temp (°C)` = col_double(),
  `Rel Hum (%)` = col_double(),
  `Wind Dir (10s deg)` = col_double(),
  `Wind Spd (km/h)` = col_double(),
  `Visibility (km)` = col_double(),
  `Stn Press (kPa)` = col_double()
)
See spec(...) for full column specifications.
 Error in attr(x, "names") <- as.character(value) : 
  'names' attribute [27] must be the same length as the vector [24] 
```


So probably that reading directly JSON file is a safer fix.  

Anyway, I'm posting it here anyway in case it can help. 

Thanks for your time.

**Comment by ConorIA** [comment]
I see @BastienFR gas already been working. I didn't have git on my work computer, but I came up with a workaround in the form of: https://www.diffchecker.com/y7e84er0 

This was lightly tested. I opted not to remove the new columns, though I would argue that we don't need the extra metadata if we can easily get that info from the station catalog.  

One edit to add: I assumed there may be some use of the `read_hcd()` function for cached data, so I opted to leave the old behaviour intact, this was probably not necessary. @BastienFR solution looks cleaner.

**Comment by gavinsimpson** [comment]
Thanks @BastienFR @ConorIA; I ended up changing quite a few little things in addition to just getting `read_hcd()` to parse the new files. I left the new columns in place, just slightly rearranged them. I agree that we can probably drop the Lat/Long and StationID variables, but I'll think on that a little. I decided to remove the functionality for reading/handling old files and thus streamlined the code a bit in doing so. User can always install an older version to get that functionality back - to that end I bump the patch version now to 0.0-2.
